tools/api-markdown-documenter/src/LintApiModel.ts

```diff
@@ -0,0 +1,308 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import {
+	ApiDocumentedItem,
+	type ApiItem,
+	ApiItemContainerMixin,
+	type ApiModel,
+} from "@microsoft/api-extractor-model";
+import type { DocInheritDocTag } from "@microsoft/tsdoc";
+
+import { defaultLoadModelOptions, loadModel, type LoadModelOptions } from "./LoadModel.js";
+import { noopLogger } from "./Logging.js";
+import { resolveSymbolicReference } from "./utilities/index.js";
+import { fail } from "node:assert";
+
+/**
+ * Linter check options.
+ *
+ * @public
+ */
+export interface LinterOptions {
+	/**
+	 * Whether or not to evaluate `{@link}` and `{@inheritDoc}` references as a part of the linting process.
+	 *
+	 * @remarks
+	 * Note: these checks do not evaluate the *syntax* of reference tags.
+	 * API-Extractor does that as a part of its model generation process.
+	 * This option is for verifying that symbolic references point to valid targets within the API model..
+	 *
+	 * @defaultValue `true`
+	 */
+	checkReferences?: boolean;
+}
+
+/**
+ * {@link lintApiModel} options.
+ */
+export interface LintApiModelOptions extends LoadModelOptions, LinterOptions {}
+
+/**
+ * {@link LintApiModelOptions} defaults.
+ */
+const defaultLintApiModelOptions: Required<Omit<LintApiModelOptions, "modelDirectoryPath">> = {
+	...defaultLoadModelOptions,
+	checkReferences: true,
+};
+
+/**
+ * Kinds of linter errors reported by {@link lintApiModel}.
+ */
+export enum LinterErrorKind {
+	/**
+	 * Indicates an `{@inheritDoc}` tag included an invalid reference.
+	 */
+	// eslint-disable-next-line unicorn/prevent-abbreviations
+	InheritDocReferenceError = "inheritDocReferenceError",
+
+	/**
+	 * Indicates a `{@link}` tag included an invalid reference.
+	 */
+	LinkReferenceError = "linkReferenceError",
+}
+
+/**
+ * A linter error found while checking the API model.
+ */
+export interface LinterErrorBase<TError extends LinterErrorKind> {
+	/**
+	 * The kind of error.
+	 */
+	readonly kind: TError;
+}
```

Comment:
Do we need the generic? I'd expect `readonly kind: LinterErrorKind` or similar would be enough and prevent something funky with... enum inheritance?

Same for `LinterResultBase`.

Reply:
One of my subsequent sets of changes ended up wanting to change this pattern anyways. I may just merge those changes forward here.

The original pattern I had in mind benefitted from being able to easily type-narrow on arbitrary error kinds. But I ended up going a different direction. Will update.